design/ip-filtering-design.md

```diff
@@ -0,0 +1,140 @@
+# IP Filtering 
+
+Status: Draft
+
+## Abstract
+
+Allow or deny traffic to services behind contour with an IP allowlist or an IP denylist.
+
+## Background
+
+There have been [long standing requests](https://github.com/projectcontour/contour/issues/62) to support some form of IP filtering.
+IP filtering is a useful tool for network segregation that is supported by other ingress controllers like [ingress-nginx](https://github.com/kubernetes/ingress-nginx/blob/87d1b8bbf28265386b339e65d8943d8c3f8582ed/docs/user-guide/nginx-configuration/annotations.md#whitelist-source-range) and [ambassador](https://www.getambassador.io/docs/edge-stack/latest/topics/running/ambassador#ip-allow-and-deny), and is a [common](https://github.com/projectcontour/contour/issues/62#issuecomment-698141472) [request](https://github.com/projectcontour/contour/issues/62#issuecomment-843193956) among the community, often listed as a [blocker](https://github.com/projectcontour/contour/issues/3693#issuecomment-1317049708) for adopting contour.
+
+## Goals
+
+- Support ip allowlists and denylists via Contour APIs.
+- Support filtering on request IP or on forwarded IP (i.e. via `X-Forwarded-For`)
+
+## Non Goals
+- Define Gateway API support for ip filtering. This is an obvious next step, but requires broader agreement.
+
+## High-Level Design
+
+The `route` section of an `HTTPProxy` will have two new fields added: `ipAllowPolicy` and `ipDenyPolicy`.
+
+Example `HTTPProxy`:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: basic
+spec:
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes:
```

Comment:
I can test this - I'm not 100% sure what happens, but I would defer to whatever envoy does here.

Reply:
based on how we program match conditions the more specific prefix will get the routing priority, so since we're using per-route filters the IP policy should be applied after the route is chosen